src/System.Private.ServiceModel/src/System/ServiceModel/UpnEndpointIdentity.cs

```diff
@@ -53,6 +57,110 @@ internal UpnEndpointIdentity(WindowsIdentity windowsIdentity)
             _upnSid = windowsIdentity.User;
             _hasUpnSidBeenComputed = true;
         }
+
+        internal override void EnsureIdentityClaim()
+        {
+            if (_windowsIdentity != null)
+            {
+                lock (_thisLock)
+                {
+                    if (_windowsIdentity != null)
+                    {
+                        Initialize(Claim.CreateUpnClaim(GetUpnFromWindowsIdentity(_windowsIdentity)));
+                        _windowsIdentity.Dispose();
+                        _windowsIdentity = null;
+                    }
+                }
+            }
+        }
+
+        string GetUpnFromWindowsIdentity(WindowsIdentity windowsIdentity)
+        {
+            string downlevelName = null;
+            string upnName = null;
+ 
+            try
+            {
+                downlevelName = windowsIdentity.Name;
+ 
+                if (IsMachineJoinedToDomain())
```

Comment:
Won't this line cause us to _always_ throw if a call to GetUpnFromWindowsIdentity is called?
Is that what we expect? 

This will mean that our behaviour between desktop and .NET Core will be different, by the way